bft/bft_interp/src/lib.rs

```diff
@@ -305,13 +321,82 @@ where
         let mut buffer: [u8; 1] = [0; 1];
-        buffer[0] = self.cells[self.head].get_value();
-        file.write_all(&buffer).map_err(|e| VMError::IOError {
-            instruction: self.current_instruction().instruction(),
-            source: e,
-        })
+        buffer[0] = self.current_cell().get_value();
+        file.write_all(&buffer)
+            .and_then(|_| file.flush())
+            .map_err(|e| VMError::IOError {
+                instruction: self.current_instruction().instruction(),
+                source: e,
+            })
+            .and(Ok(self.next_instruction()))
+    }
+
+    pub fn open_loop(&mut self) -> Result<usize, VMError> {
+        if self.current_cell().is_zero() {
+            assert!(matches!(
+                self.current_instruction(),
+                DecoratedInstruction::OpenLoop { .. }
+            ));
+            match self.current_instruction() {
+                DecoratedInstruction::OpenLoop {
+                    instruction: _,
+                    closer,
+                } => Ok(self
+                    .prog
+                    .position_to_index(closer.line(), closer.character())
+                    + 1),
+                _ => unreachable!(),
+            }
+        } else {
+            Ok(self.next_instruction())
+        }
+    }
+
+    pub fn close_loop(&mut self) -> Result<usize, VMError> {
+        assert!(matches!(
+            self.current_instruction(),
+            DecoratedInstruction::CloseLoop { .. }
+        ));
+
+        match self.current_instruction() {
+            DecoratedInstruction::CloseLoop {
+                instruction: _,
+                opener,
+            } => Ok(self
+                .prog()
+                .position_to_index(opener.line(), opener.character())),
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn interpret(
+        &mut self,
+        input: &mut impl Read,
+        output: &mut impl Write,
+    ) -> Result<(), VMError> {
+        while self.instruction_pointer < self.prog().decorated_instructions().len() {
+            self.instruction_pointer = self.interpret_current_instruction(input, output)?
+        }
+        Ok(())
+    }
+
+    fn interpret_current_instruction(
+        &mut self,
+        input: &mut impl Read,
+        output: &mut impl Write,
+    ) -> Result<usize, VMError> {
+        match self.current_instruction().instruction().instruction() {
+            RawInstruction::OpenLoop => self.open_loop(),
+            RawInstruction::CloseLoop => self.close_loop(),
+            RawInstruction::DecrementDataPointer => self.seek_left(),
+            RawInstruction::IncrementDataPointer => self.seek_right(),
+            RawInstruction::IncrementByte => self.increment_cell(),
+            RawInstruction::DecrementByte => self.decrement_cell(),
+            RawInstruction::GetByte => self.read_value(input),
+            RawInstruction::PutByte => self.write_value(output),
+        }
```

Comment:
Personally I'd not have had a separate method for interpreting a single instruction, though this isn't entirely bad.